lua_modules/liquidcrystal/i2c4bit.lua

```diff
@@ -0,0 +1,55 @@
+local I2C4bit = {}
+
+-- I2C GPIO expander bitmap
+local LCD_RS = 0
+-- local LCD_RW = 1
+local LCD_EN = 2
+local LCD_BACKLIGHT = 3
+local LCD_D4 = 4
+local LCD_D5 = 5
+local LCD_D6 = 6
+local LCD_D7 = 7
+
+local function send4bitI2C(value, cmd, bus_args, backlight)
+   local hi = bit.bor(bit.lshift(bit.band(bit.rshift(value, 4), 1), LCD_D4),
+		      bit.lshift(bit.band(bit.rshift(value, 5), 1), LCD_D5),
+		      bit.lshift(bit.band(bit.rshift(value, 6), 1), LCD_D6),
+		      bit.lshift(bit.band(bit.rshift(value, 7), 1), LCD_D7))
+   local lo = bit.bor(bit.lshift(bit.band(bit.rshift(value, 0), 1), LCD_D4),
+		      bit.lshift(bit.band(bit.rshift(value, 1), 1), LCD_D5),
+		      bit.lshift(bit.band(bit.rshift(value, 2), 1), LCD_D6),
+		      bit.lshift(bit.band(bit.rshift(value, 3), 1), LCD_D7))
+   if backlight then
+      cmd = bit.bor(cmd, bit.lshift(1, LCD_BACKLIGHT))
+   end
+   i2c.start(bus_args.id)
+   i2c.address(bus_args.id, bus_args.address, i2c.TRANSMITTER)
+   i2c.write(bus_args.id, bit.bor(hi, cmd, bit.lshift(1, LCD_EN)))
+   i2c.write(bus_args.id, bit.bor(hi, bit.band(cmd, bit.bnot(bit.lshift(1, LCD_EN)))))
+   i2c.write(bus_args.id, bit.bor(lo, cmd, bit.lshift(1, LCD_EN)))
+   i2c.write(bus_args.id, bit.bor(lo, bit.band(cmd, bit.bnot(bit.lshift(1, LCD_EN)))))
+   i2c.stop(bus_args.id)
+end
+
+I2C4bit.command = function(self, value)
+   send4bitI2C(value, 0x0, self._bus_args, self._backlight)
+end
+
+I2C4bit.write = function(self, value)
+   send4bitI2C(value, bit.lshift(1, LCD_RS), self._bus_args, self._backlight)
+end
+
+I2C4bit.backlight = function(self, on)
+   self._backlight = on
+   self._backend.command(self, 0x0)
+end
+
+I2C4bit.init = function(self)
+   i2c.setup(self._bus_args.id, self._bus_args.sda, self._bus_args.scl, self._bus_args.speed)
```

Comment:
Since there may be other peripherals on the bus, I think this should perhaps be conditional on `self._bus_args.sda` being not-`nil`?

Reply:
[`self._bus_args.sda`](https://nodemcu.readthedocs.io/en/master/modules/i2c/#i2csetup) is a pin number, so it's pretty safe to assume that it's not nil. This module barely fits into RAM, additional error checks would make it LFS-only (not sure, maybe it should be)

Reply:
I was meaning just making this `if self._bus_args.sda then i2c.setup(...) end` so that there's a mechanism to forgo bus initialization here.

Reply:
Added this feature